syncer/syncer.go

```diff
@@ -0,0 +1,285 @@
+				}
+			}()
+		case event.PeerDisconnected:
+			s.peerMap.Remove(peerID)
+		}
+	}
+}
+
+// GetSyncProgression returns progression
+func (s *syncer) GetSyncProgression() *progress.Progression {
+	return s.syncProgression.GetProgression()
+}
+
+// HasSyncPeer returns whether syncer has the peer to syncs blocks
+// return false if syncer has no peer whose latest block height doesn't exceed local height
+func (s *syncer) HasSyncPeer() bool {
+	if s.peerMap == nil {
+		return false
+	}
+
+	bestPeer := s.peerMap.BestPeer(nil)
+	header := s.blockchain.Header()
+
+	return bestPeer != nil && bestPeer.Number > header.Number
+}
+
+func (s *syncer) BulkSync(ctx context.Context, newBlockCallback func(*types.Block) bool) error {
+	localLatest := uint64(0)
+	updateLocalLatest := func() {
+		if header := s.blockchain.Header(); header != nil {
+			localLatest = header.Number
+		}
+	}
+
+	updateLocalLatest()
+
+	// Create a blockchain subscription for the sync progression and start tracking
+	s.syncProgression.StartProgression(localLatest+1, s.blockchain.SubscribeEvents())
+
+	// Stop monitoring the sync progression upon exit
+	defer s.syncProgression.StopProgression()
+
+	skipList := make(map[peer.ID]bool)
+
+	for {
+		bestPeer := s.peerMap.BestPeer(skipList)
+		if bestPeer == nil || bestPeer.Number <= localLatest {
+			break
+		}
+
+		// Set the target height
+		s.syncProgression.UpdateHighestProgression(bestPeer.Number)
+
+		lastNumber, _, err := s.bulkSyncWithPeer(bestPeer.ID, newBlockCallback)
+		if err != nil {
+			s.logger.Warn("Error ", err)
+			s.logger.Warn("failed to complete bulk sync with peer, try to next one", "peer ID", bestPeer.ID)
+		}
+
+		// if node could sync with the peer fully, then exit loop
+		if err == nil && lastNumber >= bestPeer.Number {
+			break
+		}
+
+		updateLocalLatest()
+
+		skipList[bestPeer.ID] = true
+	}
+
+	return nil
+}
+
+// WatchSync syncs block with the best peer until callback returns true
+func (s *syncer) WatchSync(ctx context.Context, callback func(*types.Block) bool) error {
+	localLatest := s.blockchain.Header().Number
+	skipList := make(map[peer.ID]bool)
+
+	// Loop until context is canceled
+	for {
+		//Wait for a new event to arrive
+		<-s.newStatusCh
+
+		// fetch local latest block
+		if header := s.blockchain.Header(); header != nil {
+			localLatest = header.Number
+		}
+
+		// pick one best peer
+		bestPeer := s.peerMap.BestPeer(skipList)
+		if bestPeer == nil {
+			// Empty skipList map if there are no best peers
+			skipList = make(map[peer.ID]bool)
+
+			continue
+		}
+
+		// if the bestPeer does not have a new block continue
+		if bestPeer.Number <= localLatest {
+			continue
+		}
+
+		// fetch block from the peer
+		lastNumber, shouldTerminate, err := s.bulkSyncWithPeer(bestPeer.ID, callback)
+		if err != nil {
+			s.logger.Warn("Error ", err)
+			s.logger.Warn("failed to complete bulk sync with peer, try to next one", "peer ID", bestPeer.ID)
+		}
+
+		if err != nil || lastNumber < bestPeer.Number {
```

Comment:
Considering there is an || (or) in this if statement, maybe the `err != nil` is redundant? It looks like the entire point is to ignore (skip) the peer if it returned a lesser height, regardless of the error from `bulkSyncWithPeer`.

Reply:
You're saying sounds correct. 
What do you think? @Aleksao998

Reply:
@dbrajovic, Make sense, you are right thanks

Reply:
fixed 2140ffc9bac11ba227bff5c0c932539ed4fb8549